src/compiler/program.ts

```diff
@@ -476,108 +478,133 @@ namespace ts {
+                    result.push(resolvedModule);
+                }
+                return result;
+            }
+            // At this point, we know at least one of the following hold:
             // - file has local declarations for ambient modules
-            // OR
             // - old program state is available
-            // OR
-            // - both of items above
-            // With this it is possible that we can tell how some module names from the initial list will be resolved
-            // without doing actual resolution (in particular if some name was resolved to ambient module).
-            // Such names should be excluded from the list of module names that will be provided to `resolveModuleNamesWorker`
-            // since we don't want to resolve them again.
-
-            // this is a list of modules for which we cannot predict resolution so they should be actually resolved
+            // With this information, we can infer some module resolutions without performing resolution.
+
+            /** An ordered list of module names for which we cannot recover the resolution. */
             let unknownModuleNames: string[];
-            // this is a list of combined results assembles from predicted and resolved results.
-            // Order in this list matches the order in the original list of module names `moduleNames` which is important
-            // so later we can split results to resolutions of modules and resolutions of module augmentations.
+            /**
+             * The indexing of elements in this list matches that of `moduleNames`.
+             *
+             * Before combining results, result[i] is in one of the following states:
+             * * undefined: needs to be recomputed,
+             * * predictedToResolveToAmbientModuleMarker: known to be an ambient module.
+             * Needs to be reset to undefined before returning,
+             * * ResolvedModuleFull instance: can be reused.
+             */
             let result: ResolvedModuleFull[];
-            // a transient placeholder that is used to mark predicted resolution in the result list
+            /** A transient placeholder used to mark predicted resolution in the result list. */
             const predictedToResolveToAmbientModuleMarker: ResolvedModuleFull = <any>{};
 
+
             for (let i = 0; i < moduleNames.length; i++) {
                 const moduleName = moduleNames[i];
-                // module name is known to be resolved to ambient module if
-                // - module name is contained in the list of ambient modules that are locally declared in the file
-                // - in the old program module name was resolved to ambient module whose declaration is in non-modified file
+                // TODO: if we want to reuse resolutions more aggressively, refine this to check for whether the
+                // text of the corresponding modulenames has changed.
+                if (file === oldSourceFile) {

```

Comment:
can you add a test for the `npm install` case:

- `file1.ts` => `import * as a from "a";`
- semantic errors
- Add new file `node_modules\a\index.d.ts`
- make an edit in `file2.ts`
- no semantic errors